presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RuntimeReorderJoinSides.java

```diff
@@ -0,0 +1,163 @@
+                if (resolvedNewLeft instanceof ExchangeNode && resolvedNewLeft.getSources().size() == 1) {
+                    newLeft = resolvedNewLeft.getSources().get(0);
+                    // The HashGenerationOptimizer will generate hashVariables and append to the output layout of the nodes following the same order. Therefore,
+                    // we use the index of the old hashVariable in the ExchangeNode output layout to retrieve the hashVariable from the new left node, and feed
+                    // it as the leftHashVariable of the swapped join node.
+                    if (swapped.getLeftHashVariable().isPresent()) {
+                        int hashVariableIndex = ((ExchangeNode) resolvedNewLeft).getPartitioningScheme().getOutputLayout().indexOf(swapped.getLeftHashVariable().get());
+                        leftHashVariable = Optional.of(resolvedNewLeft.getSources().get(0).getOutputVariables().get(hashVariableIndex));
+                    }
+                }
+
+                // Add additional localExchange if the new build side does not satisfy the partitioning conditions.
+                List<VariableReferenceExpression> buildJoinVariables = swapped.getCriteria().stream()
+                        .map(JoinNode.EquiJoinClause::getRight)
+                        .collect(toImmutableList());
+                PlanNode newRight = swapped.getRight();
+                if (needLocalExchange(swapped.getRight(), Optional.of(ImmutableSet.copyOf(buildJoinVariables)), context)) {
+                    if (getTaskConcurrency(context.getSession()) > 1) {
+                        newRight = systemPartitionedExchange(
+                                context.getIdAllocator().getNextId(),
+                                LOCAL,
+                                swapped.getRight(),
+                                buildJoinVariables,
+                                Optional.empty());
+                    }
+                    else {
+                        newRight = gatheringExchange(context.getIdAllocator().getNextId(), LOCAL, swapped.getRight());
+                    }
+                }
+
+                JoinNode newJoinNode = new JoinNode(
+                        swapped.getId(),
+                        swapped.getType(),
+                        newLeft,
+                        newRight,
+                        swapped.getCriteria(),
+                        swapped.getOutputVariables(),
+                        swapped.getFilter(),
+                        leftHashVariable,
+                        swapped.getRightHashVariable(),
+                        swapped.getDistributionType());
+
+                log.info("Probe size: " + leftOutputSizeInBytes + " is smaller than Build size: " + rightOutputSizeInBytes + " => invoke join swapping.");
+                return Result.ofPlanNode(newJoinNode);
+            }
+            else {
+                return Result.empty();
```

Comment:
To avoid this nesting i would rewrite it to something like this

```
if(Double.isNaN(leftOutputSizeInBytes) || Double.isNaN(rightOutputSizeInBytes)){
return empty
}
if(leftOutputSizeInBytes > rightOutputSizeInBytes){
return empty
}
...
```